Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp

```diff
@@ -3067,6 +3070,104 @@ auto AirIRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
     return { };
 }
 
+auto AirIRGenerator::addStructNew(uint32_t typeIndex, Vector<ExpressionType>& args, ExpressionType& result) -> PartialResult
+{
+    ASSERT(typeIndex < m_info.typeCount());
+    result = tmpForType(Type { TypeKind::Ref, Nullable::No, m_info.typeSignatures[typeIndex]->index() });
+
+    auto* patch = addPatchpoint(B3::Int64);
+    patch->clobber(RegisterSet::volatileRegistersForJSCall());
+    patch->resultConstraints.append(B3::ValueRep::reg(GPRInfo::returnValueGPR));
+
+    Vector<ConstrainedTmp, 8> patchArgs;
+    patchArgs.append(ConstrainedTmp(instanceValue(), B3::ValueRep::reg(GPRInfo::argumentGPR0)));
+    patchArgs.append(ConstrainedTmp(Tmp(GPRInfo::callFrameRegister), B3::ValueRep::reg(GPRInfo::argumentGPR1)));
+    for (unsigned i = 0; i < args.size(); ++i)
+        patchArgs.append(ConstrainedTmp(args[i], B3::ValueRep::stackArgument(i * sizeof(EncodedJSValue))));
+
+    PatchpointExceptionHandle handle = preparePatchpointForExceptions(patch, patchArgs);
+    patch->setGenerator([this, typeIndex, handle] (CCallHelpers& jit, const B3::StackmapGenerationParams& params) {
+        AllowMacroScratchRegisterUsage allowScratch(jit);
+        handle.generate(jit, params, this);
+        emitStructNewImpl(jit, typeIndex);
+    });
+
+    Tmp resultStruct = Tmp(GPRInfo::returnValueGPR);
+    emitPatchpoint(m_currentBlock, patch, Vector<Tmp, 8>::from(resultStruct), WTFMove(patchArgs));
+    append(Move, resultStruct, result);
+
+    return { };
+}
```

Comment:
This code shouldn't need to memmove all the values to the stack and then memmove them to the struct. We should just allocate a struct and store directly to it.